<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; sc-toolbox 0.11.2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom_cookietemple.css" type="text/css" />
      <link rel="stylesheet" href="_static/dark_mode_css/general.css" type="text/css" />
      <link rel="stylesheet" href="_static/dark_mode_css/dark.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/dark_mode_js/default_dark.js"></script>
        <script src="_static/dark_mode_js/theme_switcher.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributor Guide" href="contributing.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> sc-toolbox
          </a>
              <div class="version">
                0.11.2
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="readme.html">sc-toolbox</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#api">API</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-sc_toolbox.api.plot">Plots</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-sc_toolbox.api.calc">Calculations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-sc_toolbox.api.util">Utilities</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cli">CLI</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sc-toolbox">sc-toolbox</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sc-toolbox-create">create</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sc-toolbox-upgrade">upgrade</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributor Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sc-toolbox</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline"></a></h1>
<section id="api">
<h2>API<a class="headerlink" href="#api" title="Permalink to this headline"></a></h2>
<p>Import the sc-tools API as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sc_toolbox.api</span> <span class="k">as</span> <span class="nn">sct</span>
</pre></div>
</div>
<p>You can then access the respective modules like:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sct</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">cool_fancy_plot</span><span class="p">()</span>
</pre></div>
</div>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#module-sc_toolbox.api.plot" id="id1">Plots</a></p></li>
<li><p><a class="reference internal" href="#module-sc_toolbox.api.calc" id="id2">Calculations</a></p></li>
<li><p><a class="reference internal" href="#module-sc_toolbox.api.util" id="id3">Utilities</a></p></li>
</ul>
</div>
<section id="module-sc_toolbox.api.plot">
<span id="plots"></span><h3>Plots<a class="headerlink" href="#module-sc_toolbox.api.plot" title="Permalink to this headline"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.Colormaps">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">Colormaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#Colormaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.Colormaps" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>Available colormaps:</dt><dd><div class="line-block">
<div class="line">grey_red</div>
<div class="line">grey_blue</div>
<div class="line">grey_green</div>
<div class="line">gray_yellow</div>
<div class="line">grey_violet</div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.annotated_cell_type_umap">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">annotated_cell_type_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_color</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_color</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'on</span> <span class="pre">data'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Plot</span> <span class="pre">title'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#annotated_cell_type_umap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.annotated_cell_type_umap" title="Permalink to this definition"></a></dt>
<dd><p>Plots a UMAP which is colored by the primary_color, but also draws all labels on top of all clusters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object</p></li>
<li><p><strong>primary_color</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>Sequence</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Primary color to color all cells by, e.g. ‘genotype’</p></li>
<li><p><strong>cell_type_color</strong> (<em>str</em>) – Key containing all cell types, e.g. ‘cell_type’</p></li>
<li><p><strong>legend_loc</strong> (<em>str</em>) – Location of the legend (default: ‘on data’)</p></li>
<li><p><strong>legend_fontsize</strong> (<em>int</em>) – Font size of the legend (default: 8)</p></li>
<li><p><strong>title</strong> (<em>str</em>) – Title of the plot</p></li>
<li><p><strong>palette</strong> – Color</p></li>
<li><p><strong>cmap</strong> – Color map of the UMAP</p></li>
<li><p><strong>figsize</strong> – Size of the figure</p></li>
<li><p><strong>save</strong> – Path to save the plot to</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>fig and axs Matplotlib objects</p>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/annotated_cell_type_umap.png" src="_images/annotated_cell_type_umap.png" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.average_expression">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">average_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'identifier'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tab:blue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#average_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.average_expression" title="Permalink to this definition"></a></dt>
<dd><p>Draw a line plot showing the gene expression over time. Expression values are averaged by individual sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_expression</strong> – Data frame containing gene expression values</p></li>
<li><p><strong>genes</strong> – List of genes for which individual line plots will be generated</p></li>
<li><p><strong>order</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Order of x-axis labels from left to right</p></li>
<li><p><strong>id_label</strong> (<em>str</em>) – Adata column in which sample id information is stored</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>cluster</strong> (<em>str</em>) – Which clusters to plot. Select ‘all” if all clusters should be drawn.</p></li>
<li><p><strong>hue</strong> – Which value to color by</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>smooth</strong> – Set to true for smoothened line plot using polynomial regression</p></li>
<li><p><strong>rotation</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em>) – set to True to rotate x-axis labels 90 degrees</p></li>
<li><p><strong>order_smooth</strong> – If greater than 1, use numpy.polyfit to estimate a polynomial regression</p></li>
<li><p><strong>conf_int</strong> – Size of the confidence interval for the regression estimate</p></li>
<li><p><strong>scatter</strong> – Set to True to add average expression values per sample ID as dots</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
<li><p><strong>cols</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
</dl>
<dl>
<dt>Example smooth:</dt><dd><img alt="_images/average_expression_smooth.png" src="_images/average_expression_smooth.png" />
</dd>
<dt>Example raw:</dt><dd><img alt="_images/average_expression_raw.png" src="_images/average_expression_raw.png" />
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.average_expression_per_cell">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">average_expression_per_cell</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tick_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#average_expression_per_cell"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.average_expression_per_cell" title="Permalink to this definition"></a></dt>
<dd><p>Plots the average gene expression as a line plot per cell.
Ideally used when the scatter point should not be sample wise, but cell wise.
:param gene_expression: Data frame containing gene expression values
:param genes: List of genes for which individual line plots will be generated
:param order: Order of x-axis labels from left to right
:param xlabel: x-axis label
:param cluster: Which clusters to plot. Select ‘all” if all clusters should be drawn.
:param hue: Split expression values by this grouping, one line per category, will be drawn
:param figsize: Size of the figure as specified in matplotlib
:param smooth: Set to true for smoothened line plot using polynomial regression
:param rotation: Set to True to rotate x-axis labels 90 degrees
:param tick_size: Size of the ticks as specified in matplotlib
:param label_size: Size of the labels as specified in matplotlib
:param order_smooth: If greater than 1, use numpy.polyfit to estimate a polynomial regression
:param conf_int: Size of the confidence interval for the regression estimate
:param scatter: Set to True to add average expression values per sample ID as dots
:param cols: List of colors to use for line plot
:param save: Path to save the plot to</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>xlabel</strong> (<em>str</em>) – </p></li>
<li><p><strong>cluster</strong> (<em>str</em>) – </p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – </p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.average_expression_per_cluster">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">average_expression_per_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'identifier'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tick_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#average_expression_per_cluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.average_expression_per_cluster" title="Permalink to this definition"></a></dt>
<dd><p>Plots gene expression over time split by cluster identity.
One line per cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_expression</strong> – Data frame containing gene expression values</p></li>
<li><p><strong>genes</strong> – List of genes for which individual line plots will be generated</p></li>
<li><p><strong>order</strong> – Order of x-axis labels from left to right</p></li>
<li><p><strong>obs</strong> – Data frame containing meta data information</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>cluster</strong> (<em>str</em>) – Which clusters to plot. Select ‘all” if all clusters should be drawn.</p></li>
<li><p><strong>id_label</strong> (<em>str</em>) – Meta data column in which sample id information is stored</p></li>
<li><p><strong>hue</strong> – Split expression values by this grouping, one line per category will be drawn</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>smooth</strong> – Set to True for smoothened line plot using polynomial regression</p></li>
<li><p><strong>rotation</strong> – Set to True to rotate x-axis labels 90 degrees</p></li>
<li><p><strong>tick_size</strong> (<em>int</em>) – Size of the ticks as specified in matplotlib</p></li>
<li><p><strong>label_size</strong> (<em>int</em>) – Size of the labels as specified in matplotlib</p></li>
<li><p><strong>order_smooth</strong> – If greater than 1, use numpy.polyfit to estimate a polynomial regression</p></li>
<li><p><strong>conf_int</strong> – Size of the confidence interval for the regression estimate</p></li>
<li><p><strong>scatter</strong> – Set to True to add average expression values per sample ID as dots</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.average_expression_split_cluster">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">average_expression_split_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'identifier'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genotype'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tick_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#average_expression_split_cluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.average_expression_split_cluster" title="Permalink to this definition"></a></dt>
<dd><p>Plot average gene expression as line plots for multiple clusters at once.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_expression</strong> – Data frame containing gene expression values</p></li>
<li><p><strong>genes</strong> – List of genes for which individual line plots will be generated</p></li>
<li><p><strong>order</strong> – Order of x-axis labels from left to right</p></li>
<li><p><strong>id_label</strong> – Meta data column in which sample id information is stored</p></li>
<li><p><strong>xlabel</strong> – x-axis label</p></li>
<li><p><strong>hue</strong> – Split expression values by this grouping, one line per category, will be drawn</p></li>
<li><p><strong>cluster</strong> – Which clusters to plot. Select ‘all” if all clusters should be drawn.</p></li>
<li><p><strong>figsize</strong> – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>smooth</strong> – Set to True for smoothened line plot using polynomial regression</p></li>
<li><p><strong>rotation</strong> – x-axis label rotation</p></li>
<li><p><strong>cols</strong> – List of colors to use for line plot</p></li>
<li><p><strong>tick_size</strong> – Size of the ticks as specified in matplotlib</p></li>
<li><p><strong>label_size</strong> – Size of the labels as specified in matplotlib</p></li>
<li><p><strong>order_smooth</strong> – If greater than 1, numpy.polyfit is used to estimate a polynomial regression</p></li>
<li><p><strong>conf_int</strong> – Size of the confidence interval for the regression estimate</p></li>
<li><p><strong>scatter</strong> – Set to True to add average expression values per sample ID as dots</p></li>
<li><p><strong>save</strong> – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<dl>
<dt>Example smooth:</dt><dd><img alt="_images/average_expression_per_cluster_smooth.png" src="_images/average_expression_per_cluster_smooth.png" />
</dd>
<dt>Example raw:</dt><dd><img alt="_images/average_expression_per_cluster_raw.png" src="_images/average_expression_per_cluster_raw.png" />
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.cluster_composition_stacked_barplot">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">cluster_composition_stacked_barplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">relative_frequencies</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">10)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">error_bar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tick_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">capsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">margins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.02,</span> <span class="pre">0.04)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#cluster_composition_stacked_barplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.cluster_composition_stacked_barplot" title="Permalink to this definition"></a></dt>
<dd><p>Plot relative frequencies as a stacked barplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relative_frequencies</strong> (<em>pandas.core.frame.DataFrame</em>) – Data frame containing relative Frequencies as calculated by calc_relFreq()</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>width</strong> (<em>float</em>) – Width of the bars</p></li>
<li><p><strong>order</strong> – Order of x-axis labels from left to right</p></li>
<li><p><strong>error_bar</strong> – Set to True to add error bars (only possible when grouping the frequencies)</p></li>
<li><p><strong>tick_size</strong> (<em>int</em>) – Size of the ticks as specified in matplotlib</p></li>
<li><p><strong>label_size</strong> (<em>int</em>) – Size of the labels as specified in matplotlib</p></li>
<li><p><strong>capsize</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em>) – Size of the horizontal lines of the error bar</p></li>
<li><p><strong>margins</strong> (<em>Tuple</em><em>[</em><em>float</em><em>, </em><em>float</em><em>]</em>) – Change margins of the plot if desired</p></li>
<li><p><strong>colors</strong> – List of colors to use for the bands</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/cluster_composition_stacked_barplot.png" src="_images/cluster_composition_stacked_barplot.png" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.colors_overview">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">colors_overview</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">colors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#colors_overview"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.colors_overview" title="Permalink to this definition"></a></dt>
<dd><p>Draw an overview plot of all used colors</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>colors</strong> (<em>Dict</em>) – Dictionary of color name and color</p></li>
<li><p><strong>ncols</strong> (<em>int</em>) – How many columns for the plot</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/colors.png" src="_images/colors.png" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.custom_plot_size">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">custom_plot_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#custom_plot_size"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.custom_plot_size" title="Permalink to this definition"></a></dt>
<dd><p>Create a custom axis object of desired sizes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>width</strong> (<em>int</em>) – Desired plot width</p></li>
<li><p><strong>height</strong> (<em>int</em>) – Desired plot height</p></li>
<li><p><strong>dpi</strong> (<em>int</em>) – Desired plot DPI.</p></li>
</ul>
</dd>
</dl>
<p>Returns: Axis of desired sizes</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.gene_boxplot">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">gene_boxplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_types'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(10,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Axin2'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#gene_boxplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.gene_boxplot" title="Permalink to this definition"></a></dt>
<dd><p>Plot gene values as split boxplots</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> – Pandas DataFrame</p></li>
<li><p><strong>palette</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – </p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>hue</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – </p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>legend</strong> – Whether to draw a legend or not</p></li>
<li><p><strong>score</strong> – </p></li>
<li><p><strong>scatter</strong> – </p></li>
<li><p><strong>rotate</strong> – </p></li>
<li><p><strong>width</strong> – Width of the desired plot</p></li>
<li><p><strong>save</strong> – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/gene_boxplot.png" src="_images/gene_boxplot.png" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.gene_expression_dpt_ordered">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">gene_expression_dpt_ordered</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#gene_expression_dpt_ordered"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.gene_expression_dpt_ordered" title="Permalink to this definition"></a></dt>
<dd><p>Plot smoothed expression of all cells ordered by pseudo time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – AnnData object</p></li>
<li><p><strong>genes</strong> – List of genes for which individual line plots will be generated</p></li>
<li><p><strong>xlabel</strong> – x-axis label</p></li>
<li><p><strong>order</strong> – Order of x-axis labels from left to right</p></li>
<li><p><strong>conf_int</strong> – Size of the confidence interval for the regression estimate</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>condition</strong> – Split expression values by this grouping, one line per category will be drawn</p></li>
<li><p><strong>label_size</strong> (<em>int</em>) – Size of the labels as specified in matplotlib</p></li>
<li><p><strong>cols</strong> – List of colors to use for line plot</p></li>
<li><p><strong>scale</strong> – Set to True to scale expression value to a range between 0 and 1</p></li>
<li><p><strong>ylim</strong> – Upper limit on the y-axis if desired</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/gene_expression_dpt_ordered.png" src="_images/gene_expression_dpt_ordered.png" />
<dl>
<dt>Example with columns:</dt><dd><img alt="_images/gene_expression_dpt_ordered_col.png" src="_images/gene_expression_dpt_ordered_col.png" />
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.genotype_vs_genotype_umaps">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">genotype_vs_genotype_umaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_label_1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_label_2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_one_legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">6)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#genotype_vs_genotype_umaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.genotype_vs_genotype_umaps" title="Permalink to this definition"></a></dt>
<dd><p>Plots a two UMAPs of genotypes next to each other displaying only the colors of the</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object</p></li>
<li><p><strong>genotype_key</strong> (<em>str</em>) – Key of the genotypes</p></li>
<li><p><strong>genotype_label_1</strong> (<em>str</em>) – Name of the first genotype; Must be contained in the genotypes</p></li>
<li><p><strong>genotype_label_2</strong> (<em>str</em>) – Name of the second genotype; Must be contained in the genotypes</p></li>
<li><p><strong>color</strong> (<em>str</em>) – Key to color by</p></li>
<li><p><strong>hide_one_legend</strong> (<em>bool</em>) – Whether to hide the legend of the genotype_label_1</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/genotype_vs_genotype_umaps.png" src="_images/genotype_vs_genotype_umaps.png" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.marker_dendrogram">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">marker_dendrogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">marker_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_fc_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'log_FC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'top'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(10,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#marker_dendrogram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.marker_dendrogram" title="Permalink to this definition"></a></dt>
<dd><p>Plots a dendogram of used marker genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>marker_table</strong> (<em>pandas.core.frame.DataFrame</em>) – A marker table as generated by sct.calc.extended_marker_table</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – Threshold for the log fold change</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Column to create pivot by; usually just the clusters</p></li>
<li><p><strong>log_fc_key</strong> (<em>str</em>) – Key for the stored log fold changes in the marker table</p></li>
<li><p><strong>label_size</strong> (<em>int</em>) – Font size of the labels</p></li>
<li><p><strong>orientation</strong> (<em>str</em>) – Orientation of the figure; Currently just ‘top’ or no orientation</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/marker_dendrogram.png" src="_images/marker_dendrogram.png" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.relative_frequencies_boxplots">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">relative_frequencies_boxplots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">relative_frequencies</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'batch'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#relative_frequencies_boxplots"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.relative_frequencies_boxplots" title="Permalink to this definition"></a></dt>
<dd><p>Plots the relative frequencies as split boxplots.
Use calc_relative_frequencies to get the required input format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relative_frequencies</strong> (<em>pandas.core.frame.DataFrame</em>) – Calculated by calc_relative_frequencies as Pandas DataFrame</p></li>
<li><p><strong>cluster</strong> – Cluster to be plotted</p></li>
<li><p><strong>cols</strong> – List of colors to use for boxes</p></li>
<li><p><strong>order</strong> – Order of x-axis labels from left to right</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>hue</strong> (<em>str</em>) – Value to color by</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>width</strong> (<em>float</em>) – Width of the plot as specified in matplotlib</p></li>
<li><p><strong>jitter</strong> – Set to True for individual dots per sample</p></li>
<li><p><strong>save</strong> – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/relative_frequencies_boxplots.png" src="_images/relative_frequencies_boxplots.png" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.relative_frequencies_lineplot">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">relative_frequencies_lineplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">relative_frequencies</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'relative</span> <span class="pre">frequency'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tick_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#relative_frequencies_lineplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.relative_frequencies_lineplot" title="Permalink to this definition"></a></dt>
<dd><p>Plot relative frequencies as a line plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relative_frequencies</strong> (<em>pandas.core.frame.DataFrame</em>) – Data frame containing relative Frequencies as calculated by calc_relFreq()</p></li>
<li><p><strong>order</strong> – Order of x-axis labels from left to right</p></li>
<li><p><strong>cluster</strong> – Which cluster to plot</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>ylabel</strong> (<em>str</em>) – y-axis label</p></li>
<li><p><strong>hue</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Value to color by</p></li>
<li><p><strong>smooth</strong> (<em>Optional</em><em>[</em><em>bool</em><em>]</em>) – Whether to smoothen the plot</p></li>
<li><p><strong>cols</strong> – List of colors to use for line plot</p></li>
<li><p><strong>title</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Title of the plot</p></li>
<li><p><strong>rotation</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em>) – Rotation of the x-axis labels</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>tick_size</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em>) – Size of the ticks as specified in matplotlib</p></li>
<li><p><strong>label_size</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em>) – Size of the labels as specified in matplotlib</p></li>
<li><p><strong>order_smooth</strong> (<em>int</em>) – If greater than 1, numpy.polyfit is used to estimate a polynomial regression</p></li>
<li><p><strong>conf_int</strong> – Size of the confidence interval for the regression estimate</p></li>
<li><p><strong>scatter</strong> – Set to True to add average expression values per sample ID as dots</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/relative_frequencies_lineplots.png" src="_images/relative_frequencies_lineplots.png" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.split_boxplot">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">split_boxplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#split_boxplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.split_boxplot" title="Permalink to this definition"></a></dt>
<dd><p>Draws a boxsplit split by hue.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> – Table containing the data to draw the boxplots for</p></li>
<li><p><strong>order</strong> – Order of the boxplot labels</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>ylabel</strong> (<em>str</em>) – y-axis label</p></li>
<li><p><strong>column</strong> – </p></li>
<li><p><strong>hue</strong> – Value to split relative frequencies by</p></li>
<li><p><strong>cols</strong> – List of colors to use for boxes</p></li>
<li><p><strong>width</strong> (<em>float</em>) – Width of the desired plot</p></li>
<li><p><strong>title</strong> – Title of the plot</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>jitter</strong> – Set to True for individual dots per sample</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.standard_lineplot">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">standard_lineplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tick_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_smooth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confidence_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#standard_lineplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.standard_lineplot" title="Permalink to this definition"></a></dt>
<dd><p>Draws a standard line plot based on Seaborn’s lmplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – Data frame containing averaged expression values</p></li>
<li><p><strong>order</strong> (<em>List</em>) – Order of x-axis labels from left to right</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>ylabel</strong> (<em>str</em>) – y-axis label</p></li>
<li><p><strong>hue</strong> – Subsets of the data which will be drawn on separate facets in the grid. Example: “condition”</p></li>
<li><p><strong>gene</strong> – Gene of interest</p></li>
<li><p><strong>smooth</strong> (<em>Optional</em><em>[</em><em>bool</em><em>]</em>) – Whether to smoothen (interpolate) the curve</p></li>
<li><p><strong>cols</strong> – List of colors to use for line plot</p></li>
<li><p><strong>title</strong> – Title of the plot</p></li>
<li><p><strong>rotation</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em>) – Rotation of the x-axis labels</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>tick_size</strong> – Size of the ticks as specified in matplotlib</p></li>
<li><p><strong>label_size</strong> – Size of the labels as specified in matplotlib</p></li>
<li><p><strong>order_smooth</strong> (<em>int</em>) – If greater than 1, numpy.polyfit is used to estimate a polynomial regression</p></li>
<li><p><strong>confidence_interval</strong> – Confidence interval</p></li>
<li><p><strong>scatter</strong> – Set to true in order to add mean expression per sample in form of scatter point</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.plot.volcano_plot">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.plot.</span></span><span class="sig-name descname"><span class="pre">volcano_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_fc_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_p_val</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'adj_p_val'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_fc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'avg_logFC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tab:orange'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tab:blue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/plot.html#volcano_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.plot.volcano_plot" title="Permalink to this definition"></a></dt>
<dd><p>Scatter plot of differential gene expression results generated by diffxpy</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> – diffxpy generated table of results</p></li>
<li><p><strong>fdr_thresh</strong> (<em>Optional</em><em>[</em><em>float</em><em>]</em>) – -log(FDR) threshold for labeling genes. If set to None, we
will consider the 99th percentile of -log(FDR) values the threshold.</p></li>
<li><p><strong>log_fc_thresh</strong> (<em>float</em>) – absolute(log_fc) threshold for labeling genes.</p></li>
<li><p><strong>adj_p_val</strong> (<em>str</em>) – Label of the adjusted p value, these are considered FDRs</p></li>
<li><p><strong>log_fc</strong> (<em>str</em>) – Label of the log fold change</p></li>
<li><p><strong>gene</strong> (<em>str</em>) – Label of column with gene names</p></li>
<li><p><strong>col</strong> (<em>str</em>) – Color of dots</p></li>
<li><p><strong>sig_col</strong> (<em>str</em>) – Colour of dots surpassing defined FDR threshold</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the figure as specified in matplotlib</p></li>
<li><p><strong>save</strong> – Path to save the plot to</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<img alt="_images/diffxpy_volcano.png" src="_images/diffxpy_volcano.png" />
</dd></dl>

</section>
<section id="module-sc_toolbox.api.calc">
<span id="calculations"></span><h3>Calculations<a class="headerlink" href="#module-sc_toolbox.api.calc" title="Permalink to this headline"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.add_percentages">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">add_percentages</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#add_percentages"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.add_percentages" title="Permalink to this definition"></a></dt>
<dd><p>Add columns to existing diffxpy table specifying percentage of expressing cells</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object containing the data</p></li>
<li><p><strong>table</strong> – Table as generated by diffxpy</p></li>
<li><p><strong>ids</strong> – </p></li>
<li><p><strong>group_by</strong> (<em>str</em>) – Label to group by</p></li>
<li><p><strong>threshold</strong> (<em>int</em>) – </p></li>
<li><p><strong>gene_label</strong> (<em>str</em>) – Label of the genes</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Table containing percentage of expressing cells</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.automated_marker_annotation">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">automated_marker_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tissue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reference'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_fold_change</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#automated_marker_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.automated_marker_annotation" title="Permalink to this definition"></a></dt>
<dd><p>Calculates a marker gene overlap based on pre-existing annotations.</p>
<p>Currently supported marker files:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 22%" />
<col style="width: 22%" />
<col style="width: 56%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Organism</p></th>
<th class="head"><p>Tissue</p></th>
<th class="head"><p>Marker File</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Mouse</p></td>
<td><p>Lung</p></td>
<td><p>lung_particle_markers.txt</p></td>
</tr>
<tr class="row-odd"><td><p>Human</p></td>
<td><p>NA</p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata._core.anndata.AnnData</em>) – AnnData object containing ranked genes</p></li>
<li><p><strong>organism</strong> (<em>str</em>) – Currently supported: ‘mouse’</p></li>
<li><p><strong>tissue</strong> (<em>str</em>) – Currently supported: ‘lung’</p></li>
<li><p><strong>marker_file</strong> (<em>str</em>) – Name of the marker file to be used - refer to table</p></li>
<li><p><strong>key</strong> (<em>str</em>) – Key of ranked genes in adata (default: ‘rank_genes_groups’)</p></li>
<li><p><strong>normalize</strong> (<em>Optional</em><em>[</em><em>Literal</em><em>[</em><em>'reference'</em><em>, </em><em>'data'</em><em>]</em><em>]</em>) – Normalization option for the marker gene overlap output (default: ‘reference’)</p></li>
<li><p><strong>p_value</strong> (<em>float</em>) – p-value threshold for existing marker genes (default: 0.05)</p></li>
<li><p><strong>log_fold_change</strong> (<em>float</em>) – log fold change threshold for existing marker genes (default: 2)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas DataFrame of overlapping genes. Visualize with a Seaborn Heatmap</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.correlate_means_to_gene">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">correlate_means_to_gene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">means</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corr_gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'EOMES'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#correlate_means_to_gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.correlate_means_to_gene" title="Permalink to this definition"></a></dt>
<dd><p>Calculate gene to gene correlation based on a mean expression table</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>means</strong> (<em>pandas.core.frame.DataFrame</em>) – </p></li>
<li><p><strong>corr_gene</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas DataFrame of correlations</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.correlate_to_signature">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">correlate_to_signature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_fc_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AT2</span> <span class="pre">cells'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_type'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_fc_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logfoldchange'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#correlate_to_signature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.correlate_to_signature" title="Permalink to this definition"></a></dt>
<dd><p>Correlations Score (based on cell type signature (logFC)) - alternative to sc.tl.score</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object containing the data</p></li>
<li><p><strong>marker</strong> (<em>pandas.core.frame.DataFrame</em>) – Pandas DataFrame containing marker genes</p></li>
<li><p><strong>log_fc_threshold</strong> (<em>float</em>) – Log fold change label</p></li>
<li><p><strong>cell_type</strong> (<em>str</em>) – Cell type to calculate the correlation for</p></li>
<li><p><strong>cell_type_label</strong> (<em>str</em>) – Label of all cell types in the AnnData object</p></li>
<li><p><strong>log_fc_label</strong> (<em>str</em>) – Label of fold change in the AnnData object</p></li>
<li><p><strong>gene_label</strong> (<em>str</em>) – Label of genes in the AnnData object</p></li>
<li><p><strong>use_raw</strong> (<em>bool</em>) – Whether to use adata.raw.X</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of correlations</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.extended_marker_table">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">extended_marker_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qval_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_type'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ranks_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#extended_marker_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.extended_marker_table" title="Permalink to this definition"></a></dt>
<dd><p>Generates an extended marker table with cell types and percentages of expressed cell types per cluster.
Run scanpy.tl.rank_genes_groups before using this function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata._core.anndata.AnnData</em>) – AnnData object containing ranked genes</p></li>
<li><p><strong>qval_thresh</strong> (<em>float</em>) – Threshold to filter the log fold change for</p></li>
<li><p><strong>cell_type_label</strong> (<em>str</em>) – Label containing all cell types</p></li>
<li><p><strong>gene_ranks_key</strong> (<em>str</em>) – Key for the ranked gene groups (generated by sc.tl.rank_genes_groups)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Pandas DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.generate_count_object">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">generate_count_object</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'disease'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_type'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cells</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'healthy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outliers_removal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#generate_count_object"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.generate_count_object" title="Permalink to this definition"></a></dt>
<dd><p>&#64;Meshal what is this really supposed to do?</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object</p></li>
<li><p><strong>hue</strong> (<em>str</em>) – Value to color by</p></li>
<li><p><strong>cell_type_label</strong> (<em>str</em>) – Label containing cell types</p></li>
<li><p><strong>cell_type</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Cells type to generate counts for</p></li>
<li><p><strong>min_samples</strong> (<em>int</em>) – Minimum samples for outlier removal with DBScan</p></li>
<li><p><strong>min_cells</strong> (<em>int</em>) – Minimal number of cells</p></li>
<li><p><strong>ref</strong> (<em>str</em>) – </p></li>
<li><p><strong>subset</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – </p></li>
<li><p><strong>layer</strong> (<em>str</em>) – </p></li>
<li><p><strong>outliers_removal</strong> (<em>bool</em>) – Whether to remove outliers or not</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>AnnData object containing counts</p>
</dd>
</dl>
<p>Example Call:
subset = [‘3d PI-KO’, ‘3d PI-WT’]</p>
<dl class="simple">
<dt>raw_counts = generate_count_object(adata,</dt><dd><p>condition = “grouping”,
cell_type_label = “celltype_refined”, cell_type = [“AT2”],
ref = “3d PI-WT”,
subset = subset)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.generate_expression_table">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">generate_expression_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_type'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#generate_expression_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.generate_expression_table" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Anndata object</p></li>
<li><p><strong>cluster</strong> (<em>str</em>) – Which label of the subsets to generate the table for. Use ‘all’ if for all subsets.</p></li>
<li><p><strong>subset_by</strong> (<em>str</em>) – Which label to subset the clusters by</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis</p></li>
<li><p><strong>hue</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Value to color by</p></li>
<li><p><strong>use_raw</strong> (<em>Optional</em><em>[</em><em>bool</em><em>]</em>) – Whether to use adata.raw.X for the calculations</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Gene expression table</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.generate_pseudobulk">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">generate_pseudobulk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'identifier'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\t'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#generate_pseudobulk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.generate_pseudobulk" title="Permalink to this definition"></a></dt>
<dd><p>Generates a pseudobulk for a given key of groups in the AnnData object.
Looks like:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 38%" />
<col style="width: 38%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Genes</p></th>
<th class="head"><p>Group Member 1</p></th>
<th class="head"><p>Group Member 2</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Gene 1</p></td>
<td><p>Value 1</p></td>
<td><p>Value 2</p></td>
</tr>
<tr class="row-odd"><td><p>Gene 2</p></td>
<td><p>Value 2</p></td>
<td><p>Value 3</p></td>
</tr>
</tbody>
</table>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata._core.anndata.AnnData</em>) – AnnData object</p></li>
<li><p><strong>group_key</strong> (<em>str</em>) – The key to group by. E.g. by mice, by condition, … (default: ‘identifier’)</p></li>
<li><p><strong>sep</strong> – Separator to use when saving the pseudobulk table (default: ‘      ‘)</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to save the pseudobulk table to (default: None)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Pandas DataFrame containing the pseudobulk table</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.core.frame.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.ranksums_between_groups">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">ranksums_between_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bystander'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'infected'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'condition'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cells</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Axin2'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#ranksums_between_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.ranksums_between_groups" title="Permalink to this definition"></a></dt>
<dd><p>Perform Wilcoxon Rank-sum test between two groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> – </p></li>
<li><p><strong>id1</strong> (<em>str</em>) – </p></li>
<li><p><strong>id2</strong> (<em>str</em>) – </p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>cells</strong> – </p></li>
<li><p><strong>score</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas DataFrame containing test statistic and p-value</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.relative_frequencies">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">relative_frequencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_type'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'batch'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#relative_frequencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.relative_frequencies" title="Permalink to this definition"></a></dt>
<dd><p>Calculates the relative frequencies of conditions grouped by an observation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData Objet containing the data</p></li>
<li><p><strong>group_by</strong> (<em>str</em>) – </p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>condition</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Relative frequencies in a Pandas DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.relative_frequency_per_cluster">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">relative_frequency_per_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_type'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'days'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#relative_frequency_per_cluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.relative_frequency_per_cluster" title="Permalink to this definition"></a></dt>
<dd><p>Calculates relative frequencies per cluster</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object containing the data</p></li>
<li><p><strong>group_by</strong> (<em>str</em>) – The label to group by for the clusters</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – x-axis label</p></li>
<li><p><strong>condition</strong> – condition to combine by</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas DataFrame of relative frequencies</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.remove_outliers">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">remove_outliers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cords</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#remove_outliers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.remove_outliers" title="Permalink to this definition"></a></dt>
<dd><p>Remove outlying cells based on UMAP embeddings with DBScan (density based clustering)
Call as: sub.obs[“d_cluster”] = remove_outliers(sub.obsm[“X_umap”], min_samples = 10)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cords</strong> – adata UMAP coordinates, typically adata.obsm[“X_umap”]</p></li>
<li><p><strong>eps</strong> (<em>int</em>) – Maximum distance between two clusters to still be considered neighbors</p></li>
<li><p><strong>min_samples</strong> (<em>int</em>) – Minimum samples of a cluster</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas DataFrame of clusters</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.calc.tidy_de_table">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.calc.</span></span><span class="sig-name descname"><span class="pre">tidy_de_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">de_test</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cells</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qval_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'treatment'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/calc.html#tidy_de_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.calc.tidy_de_table" title="Permalink to this definition"></a></dt>
<dd><p>Sorts diffxpy de table and adds percentages of expression per group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>de_test</strong> – diffxpy de test</p></li>
<li><p><strong>adata</strong> – AnnData object</p></li>
<li><p><strong>cells</strong> – </p></li>
<li><p><strong>ids</strong> – </p></li>
<li><p><strong>qval_thresh</strong> (<em>float</em>) – </p></li>
<li><p><strong>group_by</strong> (<em>str</em>) – </p></li>
<li><p><strong>cols</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas Dataframe of diffxpy table with percentages</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-sc_toolbox.api.util">
<span id="utilities"></span><h3>Utilities<a class="headerlink" href="#module-sc_toolbox.api.util" title="Permalink to this headline"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.util.binarize_score">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.util.</span></span><span class="sig-name descname"><span class="pre">binarize_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/util.html#binarize_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.util.binarize_score" title="Permalink to this definition"></a></dt>
<dd><p>Binarizes a provided key of an AnnData object by labeling values over a threshold as ‘positive’ or ‘negative’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object to perform the binarization for</p></li>
<li><p><strong>score_label</strong> (<em>str</em>) – Label in adata.obs which will be used for the thresholding</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – The threshold to evaluate for</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of ‘positive’ for all scores &gt; threshold and ‘negative’ else</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.util.read_concatenate_to_adata">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.util.</span></span><span class="sig-name descname"><span class="pre">read_concatenate_to_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_paths</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/util.html#read_concatenate_to_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.util.read_concatenate_to_adata" title="Permalink to this definition"></a></dt>
<dd><p>Parses a list of file paths and concatenates them memory efficiently into a single AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_paths</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of file paths</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Single AnnData object containing all concatenated files</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>anndata._core.anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sc_toolbox.api.util.timestamp">
<span class="sig-prename descclassname"><span class="pre">sc_toolbox.api.util.</span></span><span class="sig-name descname"><span class="pre">timestamp</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sc_toolbox/api/util.html#timestamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sc_toolbox.api.util.timestamp" title="Permalink to this definition"></a></dt>
<dd><p>Custom timestamp in common EU format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>datetime timestamp</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="cli">
<h2>CLI<a class="headerlink" href="#cli" title="Permalink to this headline"></a></h2>
<section id="sc-toolbox">
<h3>sc-toolbox<a class="headerlink" href="#sc-toolbox" title="Permalink to this headline"></a></h3>
<p>Create state of the art projects from production ready templates.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>sc-toolbox <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sc-toolbox-v">
<span id="cmdoption-sc-toolbox-verbose"></span><span class="sig-name descname"><span class="pre">-v</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--verbose</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-sc-toolbox-v" title="Permalink to this definition"></a></dt>
<dd><p>Enable verbose output (print debug statements).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sc-toolbox-l">
<span id="cmdoption-sc-toolbox-log-file"></span><span class="sig-name descname"><span class="pre">-l</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--log-file</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;log_file&gt;</span></span><a class="headerlink" href="#cmdoption-sc-toolbox-l" title="Permalink to this definition"></a></dt>
<dd><p>Save a verbose log to a file.</p>
</dd></dl>

<section id="sc-toolbox-create">
<h4>create<a class="headerlink" href="#sc-toolbox-create" title="Permalink to this headline"></a></h4>
<p>Create a new project based on an existing template.
Usually includes a Docker container, a Conda environment and notebooks.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>sc-toolbox create <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
</section>
<section id="sc-toolbox-upgrade">
<h4>upgrade<a class="headerlink" href="#sc-toolbox-upgrade" title="Permalink to this headline"></a></h4>
<p>Checks whether the locally installed version of mlf-core is the latest.
If not pip will be invoked to upgrade mlf-core to the latest version.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>sc-toolbox upgrade <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="contributing.html" class="btn btn-neutral float-right" title="Contributor Guide" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Lukas Heumos.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>