<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" />

    <meta name="generator" content="sphinx-5.0.2, furo 2022.06.21"/>
        <title>sc_toolbox.tools - pertpy</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?digest=40978830699223671f4072448e654b5958f38b89" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/override.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/sphinx_gallery.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262per;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">pertpy</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand centered" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../usage/usage.html">Usage</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.generate_expression_table.html">sc_toolbox.tools.generate_expression_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.relative_frequencies.html">sc_toolbox.tools.relative_frequencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.relative_frequency_per_cluster.html">sc_toolbox.tools.relative_frequency_per_cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.correlate_to_signature.html">sc_toolbox.tools.correlate_to_signature</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.remove_outliers.html">sc_toolbox.tools.remove_outliers</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.add_percentages.html">sc_toolbox.tools.add_percentages</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.ranksums_between_groups.html">sc_toolbox.tools.ranksums_between_groups</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.generate_count_object.html">sc_toolbox.tools.generate_count_object</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.tidy_de_table.html">sc_toolbox.tools.tidy_de_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.correlate_means_to_gene.html">sc_toolbox.tools.correlate_means_to_gene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.extended_marker_table.html">sc_toolbox.tools.extended_marker_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.generate_pseudobulk.html">sc_toolbox.tools.generate_pseudobulk</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.automated_marker_annotation.html">sc_toolbox.tools.automated_marker_annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/tools/sc_toolbox.tools.de_res_to_anndata.html">sc_toolbox.tools.de_res_to_anndata</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.Colormaps.html">sc_toolbox.plot.Colormaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.custom_plot_size.html">sc_toolbox.plot.custom_plot_size</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.standard_lineplot.html">sc_toolbox.plot.standard_lineplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.average_expression.html">sc_toolbox.plot.average_expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.average_expression_per_cluster.html">sc_toolbox.plot.average_expression_per_cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.average_expression_split_cluster.html">sc_toolbox.plot.average_expression_split_cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.average_expression_per_cell.html">sc_toolbox.plot.average_expression_per_cell</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.gene_expression_dpt_ordered.html">sc_toolbox.plot.gene_expression_dpt_ordered</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.relative_frequencies_boxplots.html">sc_toolbox.plot.relative_frequencies_boxplots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.split_boxplot.html">sc_toolbox.plot.split_boxplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.marker_dendrogram.html">sc_toolbox.plot.marker_dendrogram</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.volcano_plot.html">sc_toolbox.plot.volcano_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.cluster_composition_stacked_barplot.html">sc_toolbox.plot.cluster_composition_stacked_barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.gene_boxplot.html">sc_toolbox.plot.gene_boxplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.colors_overview.html">sc_toolbox.plot.colors_overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.relative_frequencies_lineplot.html">sc_toolbox.plot.relative_frequencies_lineplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.annotated_cell_type_umap.html">sc_toolbox.plot.annotated_cell_type_umap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage/plot/sc_toolbox.plot.genotype_vs_genotype_umaps.html">sc_toolbox.plot.genotype_vs_genotype_umaps</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributor Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/schillerlab/sc-toolbox/discussions">Discussions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Credits</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <h1>Source code for sc_toolbox.tools</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span>

<span class="kn">from</span> <span class="nn">pandas</span> <span class="kn">import</span> <span class="n">Categorical</span>
<span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">fdrcorrection</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Literal</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">typing_extensions</span> <span class="kn">import</span> <span class="n">Literal</span>  <span class="c1"># type: ignore</span>

<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">anndata</span> <span class="kn">import</span> <span class="n">AnnData</span>
<span class="kn">from</span> <span class="nn">rich</span> <span class="kn">import</span> <span class="nb">print</span>

<span class="n">WORKING_DIRECTORY</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)</span>


<div class="viewcode-block" id="generate_expression_table"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.generate_expression_table.html#sc_toolbox.tools.generate_expression_table">[docs]</a><span class="k">def</span> <span class="nf">generate_expression_table</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">cluster</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span><span class="p">,</span>
    <span class="n">subset_by</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">xlabel</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">condition</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">use_raw</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generates a table of cells by genes of expression values as a Pandas DataFrame.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: Anndata object</span>
<span class="sd">        cluster: Which label of the subsets to generate the table for. Use &#39;all&#39; if for all subsets.</span>
<span class="sd">        subset_by: Which label to subset the clusters by</span>
<span class="sd">        xlabel: Label that will be used for subsequent line plots as x-axis label. Typically a time series such as &quot;days&quot;.</span>
<span class="sd">        condition: Column name of the condition to include.</span>
<span class="sd">        use_raw: Whether to use adata.raw.X for the calculations</span>

<span class="sd">    Returns:</span>
<span class="sd">         Gene expression table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">cluster</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
        <span class="n">cells</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cells</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span> <span class="k">if</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">cluster</span> <span class="k">else</span> <span class="kc">False</span> <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">subset_by</span><span class="p">]]</span>
    <span class="k">if</span> <span class="n">use_raw</span><span class="p">:</span>
        <span class="n">gen_expression_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">todense</span><span class="p">(),</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">var_names</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gen_expression_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">var_names</span>
        <span class="p">)</span>

    <span class="n">gen_expression_table</span><span class="p">[</span><span class="s2">&quot;identifier&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;identifier&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">xlabel</span><span class="p">:</span>
        <span class="n">gen_expression_table</span><span class="p">[</span><span class="n">xlabel</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">xlabel</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">condition</span><span class="p">:</span>
        <span class="c1"># For multiple cluster, split internally per condition</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cluster</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">subset_by</span> <span class="o">!=</span> <span class="n">condition</span><span class="p">:</span>
            <span class="n">gen_expression_table</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition</span><span class="p">],</span> <span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">subset_by</span><span class="p">])</span>
            <span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">gen_expression_table</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">gen_expression_table</span></div>


<div class="viewcode-block" id="relative_frequencies"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.relative_frequencies.html#sc_toolbox.tools.relative_frequencies">[docs]</a><span class="k">def</span> <span class="nf">relative_frequencies</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group_by</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;days&quot;</span><span class="p">,</span> <span class="n">condition</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;batch&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculates the relative frequencies of conditions grouped by an observation.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData Objet containing the data</span>
<span class="sd">        group_by: Column name to group by</span>
<span class="sd">        xlabel: x-axis label</span>
<span class="sd">        condition:</span>

<span class="sd">    Returns:</span>
<span class="sd">        Relative frequencies in a Pandas DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">freqs</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;identifier&quot;</span><span class="p">,</span> <span class="n">group_by</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;identifier&quot;</span><span class="p">])</span>
    <span class="n">ind</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">group_by</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>

    <span class="n">relative_frequencies</span> <span class="o">=</span> <span class="p">[</span><span class="n">freqs</span><span class="p">[</span><span class="n">ident</span><span class="p">]</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">freqs</span><span class="p">[</span><span class="n">ident</span><span class="p">])</span> <span class="k">for</span> <span class="n">ident</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">]</span>
    <span class="n">relative_frequencies</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">relative_frequencies</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">ind</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">samples</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># relFreqs[xlabel] = grouping.loc[samples, xlabel]  ## when using Grouping Table</span>
    <span class="n">cell_types</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">combis</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;identifier&quot;</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">])</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">combis</span><span class="p">:</span>
        <span class="n">cell_types</span><span class="p">[</span><span class="n">c</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">c</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">relative_frequencies</span><span class="p">[</span><span class="n">xlabel</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cell_types</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">relative_frequencies</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>  <span class="c1"># type: ignore</span>

    <span class="c1"># Todo, add for condition</span>
    <span class="k">if</span> <span class="n">condition</span><span class="p">:</span>
        <span class="n">combis</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;identifier&quot;</span><span class="p">,</span> <span class="n">condition</span><span class="p">])</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">combis</span><span class="p">:</span>
            <span class="n">cell_types</span><span class="p">[</span><span class="n">c</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">c</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">relative_frequencies</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cell_types</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">relative_frequencies</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>  <span class="c1"># type: ignore</span>

    <span class="k">return</span> <span class="n">relative_frequencies</span></div>


<div class="viewcode-block" id="relative_frequency_per_cluster"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.relative_frequency_per_cluster.html#sc_toolbox.tools.relative_frequency_per_cluster">[docs]</a><span class="k">def</span> <span class="nf">relative_frequency_per_cluster</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group_by</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;days&quot;</span><span class="p">,</span> <span class="n">condition</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates relative frequencies per cluster</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData object containing the data</span>
<span class="sd">        group_by: The label to group by for the clusters</span>
<span class="sd">        xlabel: x-axis label</span>
<span class="sd">        condition: condition to combine by</span>

<span class="sd">    Returns:</span>
<span class="sd">        Pandas DataFrame of relative frequencies</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">frequencies</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">group_by</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
    <span class="n">celltypes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">group_by</span><span class="p">])</span>
    <span class="n">ind</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">xlabel</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>

    <span class="n">relative_frequencies</span> <span class="o">=</span> <span class="p">[</span><span class="n">frequencies</span><span class="p">[</span><span class="n">ident</span><span class="p">]</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">frequencies</span><span class="p">[</span><span class="n">ident</span><span class="p">])</span> <span class="k">for</span> <span class="n">ident</span> <span class="ow">in</span> <span class="n">celltypes</span><span class="p">]</span>
    <span class="n">relative_frequencies</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">relative_frequencies</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">ind</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">celltypes</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">cell_types</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">combinations</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">group_by</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">])</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">combination</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">:</span>
        <span class="n">cell_types</span><span class="p">[</span><span class="n">combination</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">combination</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">relative_frequencies</span><span class="p">[</span><span class="n">group_by</span><span class="p">]</span> <span class="o">=</span> <span class="n">relative_frequencies</span><span class="o">.</span><span class="n">index</span>  <span class="c1"># type: ignore</span>

    <span class="c1"># Todo, add for condition</span>
    <span class="k">if</span> <span class="n">condition</span><span class="p">:</span>
        <span class="n">combinations</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">group_by</span><span class="p">,</span> <span class="n">condition</span><span class="p">])</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">combination</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">:</span>
            <span class="n">cell_types</span><span class="p">[</span><span class="n">combination</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">combination</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">relative_frequencies</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cell_types</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">relative_frequencies</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>  <span class="c1"># type: ignore</span>

    <span class="k">return</span> <span class="n">relative_frequencies</span></div>


<div class="viewcode-block" id="correlate_to_signature"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.correlate_to_signature.html#sc_toolbox.tools.correlate_to_signature">[docs]</a><span class="k">def</span> <span class="nf">correlate_to_signature</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">marker</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">log_fc_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
    <span class="n">cell_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;AT2 cells&quot;</span><span class="p">,</span>
    <span class="n">cell_type_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">log_fc_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;logfoldchange&quot;</span><span class="p">,</span>
    <span class="n">gene_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
    <span class="n">use_raw</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Correlations Score (based on cell type signature (logFC)) - alternative to sc.tl.score</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData object containing the data</span>
<span class="sd">        marker: Pandas DataFrame containing marker genes</span>
<span class="sd">        log_fc_threshold: Log fold change label</span>
<span class="sd">        cell_type: Cell type to calculate the correlation for</span>
<span class="sd">        cell_type_label: Label of all cell types in the AnnData object</span>
<span class="sd">        log_fc_label: Label of fold change in the AnnData object</span>
<span class="sd">        gene_label: Label of genes in the AnnData object</span>
<span class="sd">        use_raw: Whether to use adata.raw.X</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of correlations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">issparse</span>

    <span class="n">topmarker</span> <span class="o">=</span> <span class="n">marker</span><span class="p">[</span><span class="n">marker</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">cell_type_label</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span><span class="p">]</span>
    <span class="n">topmarker</span> <span class="o">=</span> <span class="n">topmarker</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">topmarker</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">log_fc_label</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">log_fc_threshold</span><span class="p">,</span> <span class="p">[</span><span class="n">gene_label</span><span class="p">,</span> <span class="n">log_fc_label</span><span class="p">]]</span>

    <span class="n">gene_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="n">topmarker</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">gene_label</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)))</span>
    <span class="n">topmarker</span> <span class="o">=</span> <span class="n">topmarker</span><span class="p">[</span><span class="n">topmarker</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">gene_label</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">gene_names</span><span class="p">)]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[bold blue]</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_names</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes used for correlation score to </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">use_raw</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
            <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="p">[:,</span> <span class="n">gene_names</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">todense</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="p">[:,</span> <span class="n">gene_names</span><span class="p">]</span><span class="o">.</span><span class="n">X</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
            <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">gene_names</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">todense</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">gene_names</span><span class="p">]</span><span class="o">.</span><span class="n">X</span>
    <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_expression</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">gene_names</span><span class="p">)</span>

    <span class="c1"># For each cell separately</span>
    <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">gene_expression</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">np</span><span class="o">.</span><span class="n">correlate</span><span class="p">(</span><span class="n">topmarker</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">log_fc_label</span><span class="p">],</span> <span class="n">gene_expression</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">c</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">gene_expression</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">res</span></div>


<div class="viewcode-block" id="remove_outliers"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.remove_outliers.html#sc_toolbox.tools.remove_outliers">[docs]</a><span class="k">def</span> <span class="nf">remove_outliers</span><span class="p">(</span><span class="n">cords</span><span class="p">,</span> <span class="n">eps</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">min_samples</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Categorical</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Remove outlying cells based on UMAP embeddings with DBScan (density based clustering).</span>

<span class="sd">    Call as: sub.obs[&quot;d_cluster&quot;] = remove_outliers(sub.obsm[&quot;X_umap&quot;], min_samples = 10)</span>

<span class="sd">    Args:</span>
<span class="sd">        cords: adata UMAP coordinates, typically adata.obsm[&quot;X_umap&quot;]</span>
<span class="sd">        eps: Maximum distance between two clusters to still be considered neighbors</span>
<span class="sd">        min_samples: Minimum samples of a cluster</span>

<span class="sd">    Returns:</span>
<span class="sd">        Pandas Categorical of clusters</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">natsort</span> <span class="kn">import</span> <span class="n">natsorted</span>
    <span class="kn">from</span> <span class="nn">sklearn.cluster</span> <span class="kn">import</span> <span class="n">DBSCAN</span>

    <span class="n">clustering</span> <span class="o">=</span> <span class="n">DBSCAN</span><span class="p">(</span><span class="n">eps</span><span class="o">=</span><span class="n">eps</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="n">min_samples</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">cords</span><span class="p">)</span>
    <span class="n">cluster</span> <span class="o">=</span> <span class="n">clustering</span><span class="o">.</span><span class="n">labels_</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;U&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">cluster</span><span class="p">,</span> <span class="n">categories</span><span class="o">=</span><span class="n">natsorted</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">cluster</span><span class="p">)))</span></div>


<div class="viewcode-block" id="add_percentages"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.add_percentages.html#sc_toolbox.tools.add_percentages">[docs]</a><span class="k">def</span> <span class="nf">add_percentages</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">table</span><span class="p">,</span> <span class="n">ids</span><span class="p">,</span> <span class="n">group_by</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">gene_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Add columns to existing diffxpy table specifying percentage of expressing cells.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData object containing the data</span>
<span class="sd">        table: Table as generated by diffxpy</span>
<span class="sd">        ids: Identifiers to add percentages for.</span>
<span class="sd">        group_by: Label to group by</span>
<span class="sd">        threshold: Cell count threshold.</span>
<span class="sd">        gene_label: Label of the genes</span>

<span class="sd">    Returns:</span>
<span class="sd">        Table containing percentage of expressing cells</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">ident</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">:</span>
        <span class="n">cells</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">group_by</span><span class="p">]</span> <span class="o">==</span> <span class="n">ident</span><span class="p">]</span>
        <span class="n">data_temp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">((</span><span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="o">/</span> <span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">gene_label</span> <span class="o">==</span> <span class="s2">&quot;index&quot;</span><span class="p">:</span>
            <span class="n">table</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pct.</span><span class="si">{</span><span class="n">ident</span><span class="si">}</span><span class="s2">s&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_temp</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">table</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">table</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pct.</span><span class="si">{</span><span class="n">ident</span><span class="si">}</span><span class="s2">s&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_temp</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">table</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">gene_label</span><span class="p">])</span><span class="o">.</span><span class="n">values</span>
    <span class="k">return</span> <span class="n">table</span></div>


<div class="viewcode-block" id="ranksums_between_groups"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.ranksums_between_groups.html#sc_toolbox.tools.ranksums_between_groups">[docs]</a><span class="k">def</span> <span class="nf">ranksums_between_groups</span><span class="p">(</span>
    <span class="n">table</span><span class="p">,</span> <span class="n">id1</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;bystander&quot;</span><span class="p">,</span> <span class="n">id2</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;infected&quot;</span><span class="p">,</span> <span class="n">xlabel</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">score</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Axin2&quot;</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform Wilcoxon Rank-sum test between two groups.</span>

<span class="sd">    Args:</span>
<span class="sd">        table:</span>
<span class="sd">        id1:</span>
<span class="sd">        id2:</span>
<span class="sd">        xlabel: x-axis label</span>
<span class="sd">        cells:</span>
<span class="sd">        score:</span>

<span class="sd">    Returns:</span>
<span class="sd">        Pandas DataFrame containing test statistic and p-value</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>

    <span class="k">if</span> <span class="n">cells</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">group1</span> <span class="o">=</span> <span class="n">table</span><span class="p">[</span><span class="n">table</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">xlabel</span><span class="p">]</span> <span class="o">==</span> <span class="n">id1</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">group2</span> <span class="o">=</span> <span class="n">table</span><span class="p">[</span><span class="n">table</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">xlabel</span><span class="p">]</span> <span class="o">==</span> <span class="n">id2</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">t</span><span class="p">,</span> <span class="n">p</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">ranksums</span><span class="p">(</span><span class="n">group1</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">score</span><span class="p">],</span> <span class="n">group2</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">score</span><span class="p">])</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;wilcoxon_ranksum&quot;</span><span class="p">,</span> <span class="s2">&quot;pval&quot;</span><span class="p">])</span>
    <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">t</span><span class="p">,</span> <span class="n">p</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="generate_count_object"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.generate_count_object.html#sc_toolbox.tools.generate_count_object">[docs]</a><span class="k">def</span> <span class="nf">generate_count_object</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">hue</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;disease&quot;</span><span class="p">,</span>
    <span class="n">cell_type_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">cell_type</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">min_samples</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">min_cells</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">ref</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;healthy&quot;</span><span class="p">,</span>
    <span class="n">subset</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">layer</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">outliers_removal</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    @Meshal what is this really supposed to do?</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData object</span>
<span class="sd">        hue: Value to color by</span>
<span class="sd">        cell_type_label: Label containing cell types</span>
<span class="sd">        cell_type: Cells type to generate counts for</span>
<span class="sd">        min_samples: Minimum samples for outlier removal with DBScan</span>
<span class="sd">        min_cells: Minimal number of cells</span>
<span class="sd">        ref:</span>
<span class="sd">        subset:</span>
<span class="sd">        layer:</span>
<span class="sd">        outliers_removal: Whether to remove outliers or not</span>

<span class="sd">    Returns:</span>
<span class="sd">        AnnData object containing counts</span>

<span class="sd">    Example Call:</span>
<span class="sd">    subset = [&#39;3d PI-KO&#39;, &#39;3d PI-WT&#39;]</span>

<span class="sd">    raw_counts = generate_count_object(adata,</span>
<span class="sd">                                       condition = &quot;grouping&quot;,</span>
<span class="sd">                                       cell_type_label = &quot;celltype_refined&quot;, cell_type = [&quot;AT2&quot;],</span>
<span class="sd">                                       ref = &quot;3d PI-WT&quot;,</span>
<span class="sd">                                       subset = subset)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">adata_subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">grouping</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">subset</span><span class="p">)]</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="p">[</span>
        <span class="kc">True</span> <span class="k">if</span> <span class="p">(</span><span class="n">adata_subset</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type_label</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="ow">in</span> <span class="n">cell_type</span><span class="p">)</span> <span class="k">else</span> <span class="kc">False</span>  <span class="c1"># type: ignore</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata_subset</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># Raw count data for diffxpy</span>
    <span class="n">obs</span> <span class="o">=</span> <span class="n">adata_subset</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">var</span> <span class="o">=</span> <span class="n">adata_subset</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">adata_raw</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">adata_subset</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="n">layer</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
    <span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">obs</span>
    <span class="n">adata_raw</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">var</span>
    <span class="n">adata_raw</span><span class="o">.</span><span class="n">obsm</span> <span class="o">=</span> <span class="n">adata_subset</span><span class="p">[</span><span class="n">cells</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Also automate tidy up with DBScan :)</span>
    <span class="k">if</span> <span class="n">outliers_removal</span><span class="p">:</span>
        <span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dcluster&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">remove_outliers</span><span class="p">(</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">],</span> <span class="n">min_samples</span><span class="o">=</span><span class="n">min_samples</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">hue</span><span class="p">,</span> <span class="s2">&quot;dcluster&quot;</span><span class="p">])</span>
        <span class="n">adata_raw</span> <span class="o">=</span> <span class="n">adata_raw</span><span class="p">[</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dcluster</span> <span class="o">==</span> <span class="s2">&quot;0&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="n">min_cells</span><span class="p">)</span>

    <span class="c1"># Set reference as first column</span>
    <span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">hue</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">reorder_categories</span><span class="p">([</span><span class="n">ref</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">setdiff1d</span><span class="p">(</span><span class="n">subset</span><span class="p">,</span> <span class="n">ref</span><span class="p">)[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">pal</span> <span class="o">=</span> <span class="n">adata_subset</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hue</span><span class="si">}</span><span class="s2">_colors&quot;</span><span class="p">]</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">hue</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">pal</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">adata_raw</span></div>


<div class="viewcode-block" id="tidy_de_table"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.tidy_de_table.html#sc_toolbox.tools.tidy_de_table">[docs]</a><span class="k">def</span> <span class="nf">tidy_de_table</span><span class="p">(</span><span class="n">de_test</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">cells</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">qval_thresh</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.9</span><span class="p">,</span> <span class="n">group_by</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;treatment&quot;</span><span class="p">,</span> <span class="n">cols</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sorts diffxpy de table and adds percentages of expression per group</span>

<span class="sd">    Args:</span>
<span class="sd">        de_test: diffxpy de test</span>
<span class="sd">        adata: AnnData object</span>
<span class="sd">        cells:</span>
<span class="sd">        ids:</span>
<span class="sd">        qval_thresh:</span>
<span class="sd">        group_by:</span>
<span class="sd">        cols:</span>

<span class="sd">    Returns:</span>
<span class="sd">        Pandas Dataframe of diffxpy table with percentages</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">de_test</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;qval&quot;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="n">result</span><span class="o">.</span><span class="n">qval</span> <span class="o">&lt;</span> <span class="n">qval_thresh</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Add percentages</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">add_percentages</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">cells</span><span class="p">],</span> <span class="n">result</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="n">ids</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="n">group_by</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="correlate_means_to_gene"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.correlate_means_to_gene.html#sc_toolbox.tools.correlate_means_to_gene">[docs]</a><span class="k">def</span> <span class="nf">correlate_means_to_gene</span><span class="p">(</span><span class="n">means</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">corr_gene</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;EOMES&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate gene to gene correlation based on a mean expression table</span>

<span class="sd">    Args:</span>
<span class="sd">        means:</span>
<span class="sd">        corr_gene:</span>

<span class="sd">    Returns:</span>
<span class="sd">        Pandas DataFrame of correlations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">scipy.stats</span>

    <span class="n">genes</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span>
    <span class="n">cors</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;spearman_corr&quot;</span><span class="p">,</span> <span class="s2">&quot;pvalue&quot;</span><span class="p">])</span>
    <span class="c1"># tab = sc.get.obs_df(sub, keys = [corr_gene], layer = None, use_raw = True)</span>
    <span class="n">table</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="n">corr_gene</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>

    <span class="c1"># Loop over all genes.</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">means</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="n">gene</span><span class="p">]])</span>  <span class="c1"># Spearman&#39;s rho</span>
        <span class="n">cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>

    <span class="n">cors</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">cors</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;spearman_corr&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">cors</span></div>


<div class="viewcode-block" id="extended_marker_table"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.extended_marker_table.html#sc_toolbox.tools.extended_marker_table">[docs]</a><span class="k">def</span> <span class="nf">extended_marker_table</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="n">qval_thresh</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">cell_type_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">gene_ranks_key</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;rank_genes_groups&quot;</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates an extended marker table with cell types and percentages of expressed cell types per cluster.</span>

<span class="sd">    Run scanpy.tl.rank_genes_groups before using this function.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData object containing ranked genes</span>
<span class="sd">        qval_thresh: Threshold to filter the log fold change for</span>
<span class="sd">        cell_type_label: Label containing all cell types</span>
<span class="sd">        gene_ranks_key: Key for the ranked gene groups (generated by sc.tl.rank_genes_groups)</span>

<span class="sd">    Returns:</span>
<span class="sd">        A Pandas DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">gene_ranks_key</span><span class="p">]</span>
    <span class="n">all_markers</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">names</span><span class="p">:</span>
        <span class="n">current</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">][</span><span class="n">cluster</span><span class="p">],</span>
                <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;scores&quot;</span><span class="p">][</span><span class="n">cluster</span><span class="p">],</span>
                <span class="s2">&quot;log_FC&quot;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">][</span><span class="n">cluster</span><span class="p">],</span>
                <span class="s2">&quot;pval&quot;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;pvals&quot;</span><span class="p">][</span><span class="n">cluster</span><span class="p">],</span>
                <span class="s2">&quot;pval_adj&quot;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;pvals_adj&quot;</span><span class="p">][</span><span class="n">cluster</span><span class="p">],</span>
                <span class="s2">&quot;cell_type&quot;</span><span class="p">:</span> <span class="n">cluster</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="p">)</span>

        <span class="c1"># Add percentage expressed per cell type</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;group&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;within&quot;</span> <span class="k">if</span> <span class="n">ct</span> <span class="o">==</span> <span class="n">cluster</span> <span class="k">else</span> <span class="s2">&quot;outside&quot;</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">cell_type_label</span><span class="p">]]</span>
        <span class="n">current</span> <span class="o">=</span> <span class="n">add_percentages</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="n">current</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="s2">&quot;group&quot;</span><span class="p">,</span> <span class="n">gene_label</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;within&quot;</span><span class="p">,</span> <span class="s2">&quot;outside&quot;</span><span class="p">])</span>

        <span class="n">all_markers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current</span><span class="p">)</span>

    <span class="n">all_markers_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">all_markers</span><span class="p">)</span>
    <span class="n">all_markers_df</span> <span class="o">=</span> <span class="n">all_markers_df</span><span class="p">[</span><span class="n">all_markers_df</span><span class="o">.</span><span class="n">pval_adj</span> <span class="o">&lt;</span> <span class="n">qval_thresh</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">all_markers_df</span></div>


<div class="viewcode-block" id="generate_pseudobulk"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.generate_pseudobulk.html#sc_toolbox.tools.generate_pseudobulk">[docs]</a><span class="k">def</span> <span class="nf">generate_pseudobulk</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span> <span class="n">group_key</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;identifier&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">save</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a pseudobulk for a given key of groups in the AnnData object.</span>

<span class="sd">    Looks like:</span>

<span class="sd">    +------------+------------------+------------------+</span>
<span class="sd">    | Genes      | Group Member 1   | Group Member 2   |</span>
<span class="sd">    +============+==================+==================+</span>
<span class="sd">    | Gene 1     | Value 1          | Value 2          |</span>
<span class="sd">    +------------+------------------+------------------+</span>
<span class="sd">    | Gene 2     | Value 2          | Value 3          |</span>
<span class="sd">    +------------+------------------+------------------+</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData object</span>
<span class="sd">        group_key: The key to group by. E.g. by mice, by condition, ... (default: &#39;identifier&#39;)</span>
<span class="sd">        sep: Separator to use when saving the pseudobulk table (default: &#39;\t&#39;)</span>
<span class="sd">        save: Path to save the pseudobulk table to (default: None)</span>

<span class="sd">    Returns:</span>
<span class="sd">        A Pandas DataFrame containing the pseudobulk table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pseudobulk</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Genes&quot;</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">group_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">:</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">group_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span>
        <span class="n">pseudobulk</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">temp</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>  <span class="c1"># column sums (genes)</span>

    <span class="k">if</span> <span class="n">save</span><span class="p">:</span>
        <span class="n">pseudobulk</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">save</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">sep</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">pseudobulk</span></div>


<div class="viewcode-block" id="automated_marker_annotation"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.automated_marker_annotation.html#sc_toolbox.tools.automated_marker_annotation">[docs]</a><span class="k">def</span> <span class="nf">automated_marker_annotation</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="n">organism</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">tissue</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">marker_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">key</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;rank_genes_groups&quot;</span><span class="p">,</span>
    <span class="n">normalize</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;reference&quot;</span><span class="p">,</span> <span class="s2">&quot;data&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="s2">&quot;reference&quot;</span><span class="p">,</span>
    <span class="n">p_value</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">log_fold_change</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculates a marker gene overlap based on pre-existing annotations.</span>

<span class="sd">    Currently supported marker files:</span>

<span class="sd">    +------------+------------+------------------------------+</span>
<span class="sd">    | Organism   | Tissue     | Marker File                  |</span>
<span class="sd">    +============+============+==============================+</span>
<span class="sd">    | Mouse      | Lung       | lung_particle_markers.txt    |</span>
<span class="sd">    +------------+------------+------------------------------+</span>
<span class="sd">    | Human      | NA         |                              |</span>
<span class="sd">    +------------+------------+------------------------------+</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData object containing ranked genes</span>
<span class="sd">        organism: Currently supported: &#39;mouse&#39;</span>
<span class="sd">        tissue: Currently supported: &#39;lung&#39;</span>
<span class="sd">        marker_file: Name of the marker file to be used - refer to table</span>
<span class="sd">        key: Key of ranked genes in adata (default: &#39;rank_genes_groups&#39;)</span>
<span class="sd">        normalize: Normalization option for the marker gene overlap output (default: &#39;reference&#39;)</span>
<span class="sd">        p_value: p-value threshold for existing marker genes (default: 0.05)</span>
<span class="sd">        log_fold_change: log fold change threshold for existing marker genes (default: 2)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Pandas DataFrame of overlapping genes. Visualize with a Seaborn Heatmap</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">supported_organisms</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;mouse&quot;</span><span class="p">}</span>
    <span class="n">supported_tissues</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;lung&quot;</span><span class="p">}</span>
    <span class="n">supported_marker_files</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;lung_particle_markers.txt&quot;</span><span class="p">}</span>

    <span class="k">if</span> <span class="n">organism</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">supported_organisms</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[bold red]Unfortunately organism </span><span class="si">{</span><span class="n">organism</span><span class="si">}</span><span class="s2"> is not yet supported.&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="k">if</span> <span class="n">tissue</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">supported_tissues</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[bold red]Unfortunately tissue </span><span class="si">{</span><span class="n">tissue</span><span class="si">}</span><span class="s2"> is not yet supported.&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="k">if</span> <span class="n">marker_file</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">supported_marker_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[bold red]Unfortunately marker file </span><span class="si">{</span><span class="n">marker_file</span><span class="si">}</span><span class="s2"> could not be found. Please check your spelling.&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="n">marker_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">WORKING_DIRECTORY</span><span class="si">}</span><span class="s2">/markers/</span><span class="si">{</span><span class="n">marker_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">marker_table</span> <span class="o">=</span> <span class="n">marker_table</span><span class="p">[</span>
        <span class="p">(</span><span class="n">marker_table</span><span class="o">.</span><span class="n">logfoldchange</span> <span class="o">&gt;</span> <span class="n">log_fold_change</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">marker_table</span><span class="o">.</span><span class="n">pval_adj</span> <span class="o">&lt;</span> <span class="n">p_value</span><span class="p">)</span>
    <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">marker</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">marker_table</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">marker_table</span><span class="p">[</span><span class="n">marker_table</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ct</span><span class="p">]</span>
        <span class="n">marker</span><span class="p">[</span><span class="n">ct</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">values</span>

    <span class="k">return</span> <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">marker_gene_overlap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">marker</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">key</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="n">normalize</span><span class="p">)</span></div>


<div class="viewcode-block" id="de_res_to_anndata"><a class="viewcode-back" href="../../usage/tools/sc_toolbox.tools.de_res_to_anndata.html#sc_toolbox.tools.de_res_to_anndata">[docs]</a><span class="k">def</span> <span class="nf">de_res_to_anndata</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="n">de_res</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="o">*</span><span class="p">,</span>
    <span class="n">groupby</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gene_id_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span>
    <span class="n">score_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;score&quot;</span><span class="p">,</span>
    <span class="n">pval_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;pvalue&quot;</span><span class="p">,</span>
    <span class="n">pval_adj_col</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">lfc_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;lfc&quot;</span><span class="p">,</span>
    <span class="n">key_added</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;rank_genes_groups&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Add a tabular differential expression result to AnnData as if it was produced by scanpy.tl.rank_genes_groups.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: Annotated data matrix</span>
<span class="sd">        de_res: Tablular DE result as Pandas DataFrame</span>
<span class="sd">        groupby: Column in `de_res` that indicates the group. This column must also exist in `adata.obs`.</span>
<span class="sd">        gene_id_col: Column in `de_res` that holds the gene identifiers</span>
<span class="sd">        score_col: Column in `de_res` that holds the score (results will be ordered by score).</span>
<span class="sd">        pval_col: Column in `de_res` that holds the unadjusted pvalue</span>
<span class="sd">        pval_adj_col: Column in `de_res` that holds the adjusted pvalue.</span>
<span class="sd">                      If not specified, the unadjusted p values will be FDR-adjusted.</span>
<span class="sd">        lfc_col: Column in `de_res` that holds the log fold change</span>
<span class="sd">        key_added: Key under which the results will be stored in adata.uns</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">groupby</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="n">groupby</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">de_res</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;groupby column must exist in both adata and de_res. &quot;</span><span class="p">)</span>
    <span class="n">res_dict</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;params&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;groupby&quot;</span><span class="p">:</span> <span class="n">groupby</span><span class="p">,</span>
            <span class="s2">&quot;reference&quot;</span><span class="p">:</span> <span class="s2">&quot;rest&quot;</span><span class="p">,</span>
            <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
            <span class="s2">&quot;use_raw&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;layer&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;corr_method&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;names&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;scores&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;pvals&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;pvals_adj&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;logfoldchanges&quot;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="p">}</span>
    <span class="n">df_groupby</span> <span class="o">=</span> <span class="n">de_res</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">groupby</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">tmp_df</span> <span class="ow">in</span> <span class="n">df_groupby</span><span class="p">:</span>
        <span class="n">tmp_df</span> <span class="o">=</span> <span class="n">tmp_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">score_col</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">res_dict</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_df</span><span class="p">[</span><span class="n">gene_id_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
        <span class="n">res_dict</span><span class="p">[</span><span class="s2">&quot;scores&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_df</span><span class="p">[</span><span class="n">score_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
        <span class="n">res_dict</span><span class="p">[</span><span class="s2">&quot;pvals&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_df</span><span class="p">[</span><span class="n">pval_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
        <span class="k">if</span> <span class="n">pval_adj_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">res_dict</span><span class="p">[</span><span class="s2">&quot;pvals_adj&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_df</span><span class="p">[</span><span class="n">pval_adj_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">res_dict</span><span class="p">[</span><span class="s2">&quot;pvals_adj&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fdrcorrection</span><span class="p">(</span><span class="n">tmp_df</span><span class="p">[</span><span class="n">pval_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>  <span class="c1"># type: ignore</span>
        <span class="n">res_dict</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_df</span><span class="p">[</span><span class="n">lfc_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>  <span class="c1"># type: ignore</span>

    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">,</span> <span class="s2">&quot;scores&quot;</span><span class="p">,</span> <span class="s2">&quot;pvals&quot;</span><span class="p">,</span> <span class="s2">&quot;pvals_adj&quot;</span><span class="p">,</span> <span class="s2">&quot;logfoldchanges&quot;</span><span class="p">]:</span>
        <span class="n">res_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span><span class="n">res_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>  <span class="c1"># type: ignore</span>
            <span class="n">columns</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">df_groupby</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">to_records</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">column_dtypes</span><span class="o">=</span><span class="s2">&quot;O&quot;</span><span class="p">)</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">key_added</span><span class="p">]</span> <span class="o">=</span> <span class="n">res_dict</span></div>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021, Lukas Heumos, Schillerlab, Theislab
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/scripts/furo.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/design-tabs.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>